p32/src/bank.rs

```diff
@@ -0,0 +1,442 @@
+#[derive(Debug, PartialEq)]
+pub struct User {
+    name: String,
+    credit_line: u64,
+    balance: i64,
+}
+
+trait BalanceAccessor {
+    fn get_balance(&self) -> i64;
+    fn set_balance(&mut self, value: i64);
+}
+
+impl BalanceAccessor for User {
+    fn get_balance(&self) -> i64 {
+        self.balance
+    }
+
+    fn set_balance(&mut self, value: i64) {
+        self.balance = value
+    }
+}
+
+impl User {
+    // type of name?
+    pub fn new(name: String, credit_line: u64, balance: i64) -> Self {
+        Self {
+            name,
+            credit_line,
+            balance,
+        }
+    }
+}
+
+impl MaxCredit for User {
+    fn max_credit(&self) -> u64 {
+        self.balance as u64 + self.credit_line
+    }
+}
+
+trait Balance {
+    fn balance(&self) -> BalanceSheet;
+}
+
+trait MaxCredit {
+    fn max_credit(&self) -> u64;
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Bank {
+    users: Vec<User>,
+    name: String,
+    credit_interest: u64,
+    debit_interest: u64,
+}
+
+impl Bank {
+    pub(crate) fn merge(&mut self, another: Bank) -> bool {
+        for user_from_new_bank in another.users {
+            let user_name = &user_from_new_bank.name;
+            match self.get_user_index_by_id(user_name) {
+                None => self.users.push(user_from_new_bank),
+                Some(overlapping_user_index) => {
+                    let old_balance = self.users[overlapping_user_index].balance;
+                    let new_balance = old_balance + user_from_new_bank.balance;
+                    self.users[overlapping_user_index].set_balance(new_balance)
+                }
+            }
+        }
+        true
+    }
+}
+
+pub struct BalanceSheet {
+    liabilities: u64,
+    assets: i64,
+}
+
+impl Balance for Bank {
+    fn balance(&self) -> BalanceSheet {
+        let total_assets = self.users.iter().map(|user| user.get_balance()).sum();
+        let total_liabilities = self.users.iter().map(|user| user.credit_line).sum();
+        BalanceSheet {
+            assets: total_assets,
+            liabilities: total_liabilities,
+        }
+    }
+}
+
+impl Bank {
+    // type of users?
+    pub fn new(users: Vec<User>, name: String, credit_interest: u64, debit_interest: u64) -> Self {
+        Self {
+            users,
+            name,
+            credit_interest,
+            debit_interest,
+        }
+    }
+    pub(crate) fn get_user_by_id(&self, user_id: String) -> Option<&User> {
+        self.users.iter().find(|&user| user.name == user_id)
+    }
+    fn get_user_index_by_id(&self, user_id: &str) -> Option<usize> {
+        self.users
+            .iter()
+            .enumerate()
+            .find(|(_, user)| user.name == user_id)
+            .map(|(a, _)| a)
```

Comment:
You could use the `position` adapter to make this code shorter: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.position